database/migrations/2024_11_19_130731_create_beatmap_versioning_tables.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        // raw statement used as laravel syntax doesn't support a fixed size BINARY column
+        DB::statement('CREATE TABLE `beatmapset_files` (
+            `file_id` BIGINT UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+            `sha2_hash` BINARY(32) NOT NULL,
+            `file_size` INT UNSIGNED NOT NULL,
+
+            UNIQUE (`sha2_hash`)
+        )');
+
+        Schema::create('beatmapset_versions', function (Blueprint $table) {
+            $table->bigIncrements('version_id');
+            $table->mediumInteger('beatmapset_id')->unsigned();
+            $table->dateTime('created_at')->default(DB::raw('NOW()'));
```

Comment:
Yeah see date types is another one of these that everything and everyone everywhere does differently. Two-minute research seems to indicate timestamp is slightly different (and probably better) mostly because it seems to be storing UTC from what I can gather from mysql docs, so will change.

Reply:
yeah the benefit is it's stored in utc and the main drawback is it'll explode in 14 years (search mysql 2038)